pkg/token/token.go

```diff
@@ -504,16 +477,103 @@ func NewVerifier(clusterID, partitionID, region string) Verifier {
 			},
 			Timeout: 10 * time.Second,
 		},
-		clusterID:         clusterID,
-		validSTShostnames: stsHostsForPartition(partitionID, region),
+		clusterID: clusterID,
+		partition: partitionID,
+		region:    region,
 	}
 }
 
-// verify a sts host, doc: http://docs.amazonaws.cn/en_us/general/latest/gr/rande.html#sts_region
+// Ensures that a given host is following the appropriate STS hostname format, e.g. sts.{region}.{suffix}.
+// Does not run any validation, so hostnames that contain invalid regions (e.g. sts.not-a-region.amazonaws.com) or
+// regions where STS/FIPS/Dualstack are not supported will be verified.
+// doc: http://docs.amazonaws.cn/en_us/general/latest/gr/rande.html#sts_region
 func (v tokenVerifier) verifyHost(host string) error {
-	if _, ok := v.validSTShostnames[host]; !ok {
+	hostRegion, err := getStsRegion(host)
+	if err != nil {
+		return FormatError{fmt.Sprintf("unexpected hostname %q in pre-signed URL", host)}
+	}
+	if hostRegion == "global" && v.partition == endpoints.AwsPartitionID {
+		return nil
+	} else if hostRegion == "global" && v.partition != endpoints.AwsPartitionID {
+		return FormatError{fmt.Sprintf("global endpoint unsupported in partition %s", v.partition)}
+	}
+	// Ensure that the region follows valid formatting, as was done in the Go SDK V1:
+	// https://github.com/aws/aws-sdk-go/blob/163aada692ed32951f979aacf452ded4c03b8a7c/aws/endpoints/v3model.go#L500
+	if !validateInputRegion(hostRegion) {
+		return FormatError{fmt.Sprintf("invalid region identifier format provided: %s", hostRegion)}
+	}
+
+	stsResolver := sts.NewDefaultEndpointResolverV2()
+
+	// Generate all possible endpoints given this region
+	var resolvedHosts []string
+	options := []bool{true, false}
+	for _, useFIPS := range options {
+		for _, useDualStack := range options {
+			// If the resolver cannot find a hostname for the given region, it will fall back to the commerical
+			// format, resulting in something like sts.{region}.amazonaws.com. So, if the hostRegion is invalid
+			// or unsupported by STS, then we expect to see an endpoint that follows the commerical partition format.
+			endpoint, err := stsResolver.ResolveEndpoint(context.TODO(), sts.EndpointParameters{
+				Region:       aws.String(hostRegion),
+				UseFIPS:      aws.Bool(useFIPS),
+				UseDualStack: aws.Bool(useDualStack),
+			})
+			if err != nil {
+				logrus.Infof("Error resolving endpoint for FIPS=%v, DualStack=%v: %v", useFIPS, useDualStack, err)
```

Comment:
We should not be logging at default level unless we are getting the endpoints for a region only once and caching because we already know this will fail in few scenarios like dualstack for iso partitions.